### PR TITLE
Yaling - L233.water.demand.livestock

### PR DESCRIPTION
**Pull request by yalingliupnnl**
#537

### CODE REVIEW

**Comment by bpbond** [comment]
Hi Yaling, hmm, this isn't passing tests. Why did you open this PR by accident?

**Comment by bpbond** [comment]
No, this doesn't pass tests for me either. Does it for you @yalingliupnnl ?

**Comment by yalingliupnnl** [comment]
@bpbond It did pass the test at my end, please refer to the latest commit.

**Comment by bpbond** [comment]
But there's no comparison data in this commit. I don't think it's possible that it's passing tests.

**Comment by yalingliupnnl** [comment]
@bpbond it does have a comparison file L133.water_demand_livestock_R_C_W_km3_Mt.csv in the directory `workspace\Github\gcamdata\inst\extdata\water`.

**Comment by bpbond** [comment]
@yalingliupnnl you seem to have forgotten to commit this file. Please add it.

**Comment by yalingliupnnl** [comment]
@bpbond I though previous issue (test package was passed at my end but not yours) is because I did not sync my local changes to the pull request. I just updated the comparison file (standardize the comments) and synced the changes.

File: R/zchunk_L233.water.demand.livestock.R [hunk 1]
```diff
@@ -1,20 +1,21 @@
 #' module_water_L233.water.demand.livestock
 #'
-#' Briefly describe what this chunk does.
+#' This chunk is to generate water coefficients for region-specific livestock for model years.
```
Comment (bpbond) [inline R/zchunk_L233.water.demand.livestock.R]: "Generate..."

File: R/zchunk_L233.water.demand.livestock.R [hunk 2]
```diff
@@ -1,20 +1,21 @@
...
 #'
 #' @param command API command to execute
 #' @param ... other optional parameters, depending on command
 #' @return Depends on \code{command}: either a vector of required inputs,
 #' a vector of output names, or (if \code{command} is "MAKE") all
 #' the generated outputs: \code{L233.TechCoef}. The corresponding file in the
 #' original data system was \code{L233.water.demand.livestock.R} (water level2).
-#' @details Describe in detail what this chunk does.
+#' @details This chunk is to generate water coefficients for region-specific livestock for
```
Comment (bpbond) [inline R/zchunk_L233.water.demand.livestock.R]: "This chunk generates..."

Please define "water coefficients" more precisely. What does this mean?

File: R/zchunk_L233.water.demand.livestock.R [hunk 3]
```diff
@@ -34,37 +35,34 @@ module_water_L233.water.demand.livestock_DISABLED <- function(command, ...) {
...
-      add_comments("comments describing how data generated") %>%
+    results %>%
+      add_title("water coefficient for region-specific livestock") %>%
+      add_units("NA") %>%
+      add_comments("The data is generated through:
+                   1) read in the file L133.water_demand_livestock_R_C_W_km3_Mt and left join with the file A_an_technology;
+                   2) generate new fields water_sector and minicam.energy.input:
+                   3) left join with GCAM_region_names;
+                   4) replicate the water coefficients for all MODEL_YEARS") %>%
       add_comments("can be multiple lines") %>%
```
Comment (bpbond) [inline R/zchunk_L233.water.demand.livestock.R]: remove

File: R/zchunk_L233.water.demand.livestock.R [hunk 4]
```diff
@@ -34,37 +35,34 @@ module_water_L233.water.demand.livestock_DISABLED <- function(command, ...) {
 
     # ===================================================
     # TRANSLATED PROCESSING CODE GOES HERE...
```
Comment (bpbond) [inline R/zchunk_L233.water.demand.livestock.R]: Remove

File: R/zchunk_L233.water.demand.livestock.R [hunk 5]
```diff
@@ -34,37 +35,34 @@ module_water_L233.water.demand.livestock_DISABLED <- function(command, ...) {
...
+    L133.water_demand_livestock_R_C_W_km3_Mt %>%
+      rename(supplysector = GCAM_commodity) %>%
+      inner_join(select(A_an_technology, supplysector, subsector, technology), by = "supplysector") %>%
+      mutate(water_sector = "Livestock") %>%
+      mutate(minicam.energy.input = set_water_input_name(water_sector, water_type, A03.sector)) %>%
+      left_join(GCAM_region_names, by = "GCAM_region_ID") %>%
+      mutate(market.name = region) -> L233.TechCoef_tmp
+      # Set the coef for all years
+    L233.TechCoef_tmp[rep(1:nrow(L233.TechCoef_tmp), times = length(MODEL_YEARS)), ] %>%
+      mutate(year = MODEL_YEARS[sort(rep(1:length(MODEL_YEARS), times = nrow(L233.TechCoef_tmp)))]) %>%
```
Comment (bpbond) [inline R/zchunk_L233.water.demand.livestock.R]: These two lines can be replaced by the MUCH simpler
```R
      repeat_add_columns(tibble(year = MODEL_YEARS)) %>%
```

I'll make and commit this change.

**Comment by yalingliupnnl** [comment]
@bpbond I addressed your comments but failed to sync the changes to the pull request, could you help me with that?

**Comment by bpbond** [comment]
@yalingliupnnl I'm out of the office all this week and next, sorry, so probably easier to see if someone like Abigail, Russell, or Aaron is available to help about the sync problem.

**Comment by yalingliupnnl** [comment]
@bpbond Sean helped me address the sync problem, now I have synced the local changes to the remote repository.

File: R/zchunk_L233.water.demand.livestock.R [hunk 6]
```diff
@@ -1,20 +1,21 @@
 #' module_water_L233.water.demand.livestock
 #'
-#' Briefly describe what this chunk does.
+#' This chunk generates water coefficients for region-specific livestock for model years.
```
Comment (bpbond) [inline R/zchunk_L233.water.demand.livestock.R]: "Generate water coefficients..."

File: R/zchunk_L233.water.demand.livestock.R [hunk 7]
```diff
@@ -32,39 +33,32 @@ module_water_L233.water.demand.livestock_DISABLED <- function(command, ...) {
...
+      add_comments("The data is generated through:
+                   1) read in the file L133.water_demand_livestock_R_C_W_km3_Mt and left join with the file A_an_technology;
+                   2) generate new fields water_sector and minicam.energy.input:
+                   3) left join with GCAM_region_names;
+                   4) replicate the water coefficients for all MODEL_YEARS") %>%
       add_legacy_name("L233.TechCoef") %>%
-      add_precursors("precursor1", "precursor2", "etc") %>%
+      add_precursors("common/GCAM_region_names", "water/A03.sector", "aglu/A_an_technology",
+                     "temp-data-inject/L133.water_demand_livestock_R_C_W_km3_Mt") %>%
       # typical flags, but there are others--see `constants.R`
```
Comment (bpbond) [inline R/zchunk_L233.water.demand.livestock.R]: Remove

File: R/zchunk_L233.water.demand.livestock.R [hunk 8]
```diff
@@ -32,39 +33,32 @@ module_water_L233.water.demand.livestock_DISABLED <- function(command, ...) {
...
+    L133.water_demand_livestock_R_C_W_km3_Mt %>%
+      rename(supplysector = GCAM_commodity) %>%
+      inner_join(select(A_an_technology, supplysector, subsector, technology), by = "supplysector") %>%
+      mutate(water_sector = "Livestock") %>%
+      mutate(minicam.energy.input = set_water_input_name(water_sector, water_type, A03.sector)) %>%
+      left_join(GCAM_region_names, by = "GCAM_region_ID") %>%
+      mutate(market.name = region) %>%
+      # Set the coef for all years
+      repeat_add_columns(tibble(year = MODEL_YEARS)) %>%
+      select(-GCAM_region_ID, -water_type, -water_sector) %>%
```
Comment (bpbond) [inline R/zchunk_L233.water.demand.livestock.R]: This can't work, can it? Please check (build and reload) before committing.

**Comment by yalingliupnnl** [comment]
@bpbond Comments are addressed and test package passed

**Comment by bpbond** [comment]
This still isn't passing tests, either on Travis or my computer.

**Comment by yalingliupnnl** [comment]
@bpbond It did pass the test at my end, could you help to take a look at it on my laptop?

**Comment by bpbond** [comment]
Yaling, it's not possible that this is passing tests on your system. Here's the beginning of the comparison file:
```
region,supplysector,subsector,technology,year,minicam.energy.input,coefficient,market.name
USA,SheepGoat,Pastoral,Scavenging_Other,1975,water_td_an_W,0.228718325623527,USA
USA,SheepGoat,Pastoral,Pasture_FodderGrass,1975,water_td_an_C,0.228718325623527,USA
```

And here's the output file from your chunk:

```
# FLAG_LONG_YEAR_FORM FLAG_NO_XYEAR
# The data is generated through:                    1) read in the file L133.water_demand_livestock_R_C_W_km3_Mt and left join with the file A_an_technology;                    2) generate new fields water_sector and minicam.energy.input:                    3) left join with GCAM_region_names;                    4) replicate the water coefficients for all MODEL_YEARS
supplysector,coefficient,subsector,technology,minicam.energy.input,region,market.name,year
Beef,0.091778589370333,Mixed,FodderHerb_Residue,water_td_an_C,USA,USA,1975
Beef,0.091778589370333,Mixed,FodderHerb_Residue,water_td_an_C,USA,USA,1990
```

Notice the two flags at the top: LONG_YEAR_FORM tells the test system to reshape these data to wide form, and NO_XYEAR tells it to add X's to the resulting year column names. But _this is not needed_ because the comparison data aren't in wide form, and _this is not possible_ because there's no `year` and `value` columns in your output. And indeed, this is exactly the error that the test system produces:

```
1. Failure: matches old data system output (@test_oldnew.R#60) -----------------
all(c("year", "value") %in% names(newdata)) isn't true.
FLAG_LONG_YEAR_FORM specified in L233.TechCoef.csv but no 'year' and 'value' columns present
```

**Comment by yalingliupnnl** [comment]
@bpbond I removed `FLAG_LONG_YEAR_FORM FLAG_NO_XYEAR` and it passed the test package at my end.

**Comment by bpbond** [comment]
@skim301 @mihejazi This PR has been open 17 days. Could someone review it please? Thanks.

File: R/zchunk_L233.water.demand.livestock.R [hunk 9]
```diff
@@ -26,47 +27,42 @@ module_water_L233.water.demand.livestock_DISABLED <- function(command, ...) {
...
+      rename(supplysector = GCAM_commodity) %>%
+      inner_join(select(A_an_technology, supplysector, subsector, technology), by = "supplysector") %>%
+      mutate(water_sector = "Livestock") %>%
+      mutate(minicam.energy.input = set_water_input_name(water_sector, water_type, A03.sector)) %>%
+      left_join(GCAM_region_names, by = "GCAM_region_ID") %>%
+      mutate(market.name = region) %>%
+      # Set the coef for all years
+      repeat_add_columns(tibble(year = MODEL_YEARS)) %>%
+      select(-GCAM_region_ID, -water_type, -water_sector) ->
+      results
```
Comment (skim301) [inline R/zchunk_L233.water.demand.livestock.R]: Please rename the "results" table to some more descriptive

**Comment by bpbond** [comment]
Thanks Sonny! @yalingliupnnl if you can make these final changes, we'll get this MERGED. 🎉

**Comment by bpbond** [comment]
Actually, this is such a trivial change I'll do it and then merge.

**Comment by bpbond** [comment]
Thank you Yaling!

**Pull request merged**